Bayesian Linear Regression now uses jaspResults

TimKDJ opened the pull request:
So we'll probably want to change the status handling and possibly the way the model is passed around.
Right now I do it "R-like", meaning I return and then pass it on explicitly. We could also use jaspResults[["stateModel"]] directly in the analyses. Thoughts?

vandenman commented on JASP-Engine/JASP/R/regressionlinearbayesian.R:
```diff
@@ -15,587 +15,360 @@
+  modelComparison$dependOnOptions(c("bayesFactorType", "bayesFactorOrder", "shownModels", "numShownModels"))
+  
+  
+  # modelComparison$addCitation <- list(
```
It's a bit ugly but you can add citations individually:
```r
modelComparison$addCitation("Clyde, M. A. (2017). BAS: Bayesian Adaptive Sampling for Bayesian Model Averaging. (Version 1.4.7)[Computer software].")
modelComparison$addCitation("Clyde, M. A., Ghosh, J., & Littman, M. L. (2011). Bayesian adaptive sampling for variable selection and model averaging. Journal of Computational and Graphical Statistics, 20, 80-101.")
```

TimKDJ replied:
Guess this method requires some work too ;)

vandenman commented on JASP-Engine/JASP/R/regressionlinearbayesian.R:
```diff
@@ -15,587 +15,360 @@
+  }
+  
+  if (sum(basregModel$nuisanceTerms) > 0) {
+    footnote <- paste("All models include ", paste(names(which(basregModel$nuisanceTerms)), collapse = ", "), ".", sep = "")
```
`paste0`?

vandenman commented on JASP-Engine/JASP/R/regressionlinearbayesian.R:
```diff
@@ -15,587 +15,360 @@
+  nuisanceTerms <- basregModel$nuisanceTerms
+  null.model <- "Null model"
+  if (sum(nuisanceTerms) > 0) {
+    null.model <- paste("Null model (incl. ", paste(names(which(nuisanceTerms)), collapse = ", "), ")", sep = "")
```
`paste0`

vandenman commented on JASP-Engine/JASP/R/regressionlinearbayesian.R:
```diff
@@ -15,587 +15,360 @@
+  
+  # generate all model names
   allModelsVisited <- any(lengths(models) == 0) # analysis will chrash if TRUE
   model.names <- vector("character", length(models))
```
`character(length(models))`?

vandenman commented on JASP-Engine/JASP/R/regressionlinearbayesian.R:
```diff
@@ -611,1138 +384,1065 @@ RegressionLinearBayesian <- function (
+  mostComplex <- which.max(lengths(basregModel$which)[topm])
+  
+  for (i in loopIdx) {
+    coefficient <- .clean(coefficients[i])
```
same thing about `.clean`

TimKDJ replied:
Sorry?

vandenman replied:
It seems some comment disappeared. If I'm not mistaken, `.clean` is done automagically by jaspResults once you enter the data.

vandenman replied:
also Joris just merged the error property PR so you could also try that out already

TimKDJ replied:
I tested and .clean() is still necessary. For example `NULL` is passed on in json as `null` which javascript cannot deal with. I would also venture a guess that it also doesn't show a nice infinity symbol when `Inf` is passed.

I added the error handling. It seems to work nicely (nice job!), I have two  "issues" with it though:
1) `basReg$getError()` does not convey its meaning super nicely.  I guess this is kinda personal, something like`basReg$hasError()` or `basReg$error()` is clearer, but I understand why you named it this way -- to match the `setError()`.
2) we HAVE to use containers now to make it work, only the containers are always displayed as such in the output and this adds a visible grouping I don't think we really need. Also, and this is something you also encountered, you cannot use this container grouping to pass dependencies from the parent to the child, instead all children MUST have some sort of dependency set otherwise they get thrown out.

vandenman commented on JASP-Engine/JASP/R/regressionlinearbayesian.R:
```diff
@@ -611,1138 +384,1065 @@ RegressionLinearBayesian <- function (
-    }
+.basregFillPlotInclusionProbabilities <- function(basregModel, jaspResults, options) {
+  
+  p <- try({
```
In this and the other `try({...})` blocks the first 20 lines aren't really prone to erroring. Perhaps it's more clear to only put things in `try({...})` that can give an error?

vandenman commented on JASP-Engine/JASP/R/regressionlinearbayesian.R:
```diff
@@ -611,1138 +384,1065 @@ RegressionLinearBayesian <- function (
+  
+  if (isTryError(p)) {
+    errorMessage <- paste("Plotting not possible:", .extractErrorMessage(p))
+    jaspResults[["inclusionProbabilitiesPlot"]]$error <- "badData" #FIXME: this should go
```
In these (and other functions) why not pass the plotObject, rather than jaspResults? Should work exactly the same but is a bit more concise.

TimKDJ replied:
Yeah, this is indeed much nicer

vandenman commented on JASP-Engine/JASP/R/regressionlinearbayesian.R:
```diff
@@ -15,587 +15,360 @@
+  .basregPlotQQ(basregModel, jaspResults, options, status)
+  
+  # this function still needs to be rewritten to ggplot (start in .plotImage.basReg):
+  .basregPlotPosteriorLogOdds(basregModel, jaspResults, options, status)
```
This means there will be a hard crash (not an R crash) when a user tries to copy this plot as pdf or eps. So for the next release this should be fixed!

TimKDJ replied:
Mkay. Maybe we can just omit the plot if no one gets it anyway...

boutinb commented:
Is this OK now?

TimKDJ commented:
Nope.. First Don's PR needs to go through so I can incorporate that stuff.

TimKDJ commented:
@vandenman maybe you can help me out here. In the code there is a warning that states
```
paste("Prior model probabilities for models with interaction effects",
                                   "are obtained from a Bernoulli (p = 0.5) prior.",
                                   "We advice using a different model prior, or excluding interaction effects.")
```
This warning is included for `(options$modelPrior == "beta.binomial")` -- is this correct? Beta binomial is apparently a Bernoulli for n=1, but is that the case here? I'm going to guess it should also be included for the `(options$modelPrior == "Bernoulli")`, given that p=0.5?

vandenman commented:
BAS doesn't give this warning anymore, so maybe we can delete it? I'm not sure how the model probabilities should be computed for a beta-binomial with interactions, but this approach is not super weird. Perhaps @AlexanderLyNL has another idea?

AlexanderLyNL commented:
Let's discuss this in person, and get to root of why bas has this warning in the first case, then we can what changed in the meantime.

vandenman commented:
@AlexanderLyNL BAS does **not** have this warning. I think it did in the past and we just copied that. But sure let's discuss this when I'm back.

TimKDJ commented:
- I removed the warning, considering BAS doesn't have it anymore
- I changed the "limit number of models shown" to strictly enforce the specified number and not go num models + 1 (for null) as per Maarten's confusion
- I made the analysis use the nifty results-merging that Joris added

vandenman commented:
@TimKDJ great! Let me know if you want me to review it again.

TimKDJ commented:
Yeah, that'd be nice!

vandenman commented on JASP-Engine/JASP/R/regressionlinearbayesian.R:
```diff
@@ -15,1771 +15,1252 @@
-
-	return(plots)
+  if (options$samplingMethod == "MCMC") {
+    postLogOddsPlot$errorMessage <- "Cannot display Posterior Log Odds when sampling method is MCMC."
```
```suggestion
    postLogOddsPlot$setError("Cannot display Posterior Log Odds when sampling method is MCMC.")
```
